From 2364c1bfd21cf47a571ada087c52e3103cdfb4da Mon Sep 17 00:00:00 2001
From: Dan Liang <dan.liang@intel.com>
Date: Thu, 2 May 2013 16:45:24 +0800
Subject: [PORT FROM MAIN]  make one xvid clip work normal in gallery

BZ: 83021

root cause: the first video sample with MPEG4 codec is a fake sample, which size
is 0, we just skip this kind of video sample, the original design can't handle this
case well.

Change-Id: I22b5a1c4d090b49b901f3cd07e1467a86c801429
Orig-Change-Id: I40af5bdff95ac2a4ee36b8f403d842ad216178e5
Signed-off-by: Gu, Wangyi <wangyi.gu@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
---
 media/libstagefright/AVIExtractor.cpp |   21 +++++++++++++++++----
 1 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/media/libstagefright/AVIExtractor.cpp b/media/libstagefright/AVIExtractor.cpp
index c4c79c3..da4fe8e 100644
--- a/media/libstagefright/AVIExtractor.cpp
+++ b/media/libstagefright/AVIExtractor.cpp
@@ -1018,11 +1018,24 @@ status_t AVIExtractor::addMPEG4CodecSpecificData(size_t trackIndex) {
     size_t size;
     bool isKey;
     int64_t timeUs;
-    status_t err =
-        getSampleInfo(trackIndex, 0, &offset, &size, &isKey, &timeUs);
 
-    if (err != OK) {
-        return err;
+    // Extract codec specific data from the first non-empty sample.
+
+    size_t sampleIndex = 0;
+    for (;;) {
+        status_t err =
+            getSampleInfo(
+                    trackIndex, sampleIndex, &offset, &size, &isKey, &timeUs);
+
+        if (err != OK) {
+            return err;
+        }
+
+        if (size > 0) {
+            break;
+        }
+
+        ++sampleIndex;
     }
 
     sp<ABuffer> buffer = new ABuffer(size);
-- 
1.7.4.1

